lib/concurrent-ruby/concurrent/collection/ruby_timeout_queue.rb

```diff
@@ -0,0 +1,58 @@
+module Concurrent
+  module Collection
+    # @!visibility private
+    # @!macro ruby_timeout_queue
+    class RubyTimeoutQueue < ::Queue
+      def initialize(*args)
+        if RUBY_VERSION >= '3.2'
+          raise "#{self.class.name} is not needed on Ruby 3.2 or later, use ::Queue instead"
+        end
+
+        super(*args)
+
+        @mutex = Mutex.new
+        @cond_var = ConditionVariable.new
+      end
+
+      def push(obj)
+        @mutex.synchronize do
+          super(obj)
+          @cond_var.signal
+        end
+      end
+      alias_method :enq, :push
+      alias_method :<<, :push
+
+      def pop(non_block = false, timeout: nil)
+        if non_block && timeout
+          raise ArgumentError, "can't set a timeout if non_block is enabled"
+        end
+
+        if non_block
+          super(true)
+        elsif @mutex.synchronize { empty? && timed_out?(timeout) { @cond_var.wait(@mutex, timeout) } }
```

Comment:
Just flagging that there could be a spurious wake up signal on the Condition Variable that prematurely invokes this condition. If that happened, I think you'd accidentally end up in a blocking pop. 

I think the recommended design is to set an additional variable flag in addition to the signal to disambiguate between the meaningful signal and the spurious ones. The primitive for this is a Concurrent::Event.

Reply:
Ah good catch, I'll address, thanks!

Reply:
> I think the recommended design is to set an additional variable flag in addition to the signal (in push) to disambiguate between the meaningful signal and the spurious ones. The primitive for this is a Concurrent::Event.

I think we can rely on `#empty?` instead? - it can only be `true` if there was a spurious wakeup.

Reply:
I've addressed, let me know what you think.

Reply:
I imagine that's sufficient 👍🏻 Thank you!

I think the only thing is there could be a race condition on `empty?` between a non-timeout `pop` and a timeout `pop` on the same Queue instance, because the non-timeout call to `super` is not synchronized with the same mutex. ...but we're building this for a very specific use case where that wouldn't happen.

Reply:
Good point! Although I think it’s best to ensure the implementation is robust in case it gets relied on more heavily in future work.

I've pushed a slightly different approach where we rely on the non-blocking super call instead.